src/fu-security-attr.c

```diff
@@ -301,17 +305,269 @@ fu_security_attrs_to_json(FuSecurityAttrs *attrs, JsonBuilder *builder)
+	if (flag_array != NULL) {
+		json_builder_set_member_name(result_builder, FWUPD_RESULT_KEY_FLAGS);
+		json_builder_begin_array(result_builder);
+		for (guint i = 0; i < flag_array->len; i++) {
+			json_builder_add_string_value(result_builder,
+						      g_strdup(g_ptr_array_index(flag_array, i)));
+		}
+		json_builder_end_array(result_builder);
+	}
+	json_builder_end_object(result_builder);
+	json_builder_end_object(result_builder);
+
+	return FALSE;
+}
+
+static void
+fu_security_attr_append_remove_to_result(JsonObject *previous_json_obj, JsonBuilder *result_builder)
+{
+	json_builder_set_member_name(
+	    result_builder,
+	    json_object_get_string_member(previous_json_obj, FWUPD_RESULT_KEY_APPSTREAM_ID));
+	json_builder_begin_object(result_builder);
+	json_builder_set_member_name(result_builder, "removed");
+	json_builder_begin_object(result_builder);
+	fu_security_attr_dup_json(previous_json_obj, result_builder);
+	json_builder_end_object(result_builder);
+	json_builder_end_object(result_builder);
+}
+
+gchar *
+fu_security_attrs_hsi_change(FuSecurityAttrs *attrs, const gchar *last_hsi_detail)
+{
+	g_autofree gchar *data = NULL;
+	g_autoptr(JsonParser) parser = json_parser_new();
+	g_autoptr(JsonGenerator) json_generator = NULL;
+	g_autoptr(JsonBuilder) result_builder = json_builder_new();
+	g_autoptr(JsonNode) result_json_root = NULL;
+	g_autoptr(GPtrArray) items = NULL;
+	g_autoptr(GHashTable) found_list = g_hash_table_new(g_str_hash, g_str_equal);
+	g_autoptr(GList) member_list = NULL;
+	g_autoptr(GList) removed_keys = NULL;
+	JsonNode *json_root = NULL;
+	JsonObject *json_obj = NULL;
+	JsonObject *previous_security_attrs = NULL;
+
+	json_parser_load_from_data(parser, last_hsi_detail, -1, NULL);
+	json_root = json_parser_get_root(parser);
+	json_obj = json_node_get_object(json_root);
+	previous_security_attrs = json_object_get_object_member(json_obj, "SecurityAttributes");
+
+	member_list = json_object_get_members(previous_security_attrs);
+	for (GList *tmp = member_list; tmp != NULL; tmp = tmp->next) {
+		g_hash_table_insert(found_list, g_strdup(tmp->data), NULL);
+	}
+
+	items = fu_security_attrs_get_all(attrs);
+	json_builder_begin_object(result_builder);
+	for (guint i = 0; i < items->len; i++) {
+		FwupdSecurityAttr *attr = g_ptr_array_index(items, i);
+		if (json_object_has_member(previous_security_attrs,
+					   fwupd_security_attr_get_appstream_id(attr)) == TRUE) {
+			/* Hit */
+			g_hash_table_remove(found_list, fwupd_security_attr_get_appstream_id(attr));
+			fu_security_attr_deep_object_compare(
+			    attr,
+			    json_object_get_object_member(
+				previous_security_attrs,
+				fwupd_security_attr_get_appstream_id(attr)),
+			    result_builder);
+		} else {
+			/* Miss- A new AppStreamID */
+			fu_security_attr_deep_object_compare(attr, NULL, result_builder);
+		}
+	}
+	removed_keys = g_hash_table_get_keys(found_list);
+	if (removed_keys != NULL) {
+		/* Removed from current */
```

Comment:
single line trivial comments like this should be lowercase please

Reply:
ok